src/terminal/adapter/MacroBuffer.cpp

```diff
@@ -0,0 +1,271 @@
+    // range of hex pairs between two `;` characters. When parsing the repeat
+    // count, we use the ExpectingRepeatCount state, but when parsing the hex
+    // pairs of the repeat, we just use the regular ExpectingHexDigit states.
+
+    auto success = true;
+    switch (_parseState)
+    {
+    case State::ExpectingText:
+        success = _appendToActiveMacro(ch);
+        break;
+    case State::ExpectingHexDigit:
+        if (_decodeHexDigit(ch))
+        {
+            _parseState = State::ExpectingSecondHexDigit;
+        }
+        else if (ch == L'!' && !_repeatPending)
+        {
+            _parseState = State::ExpectingRepeatCount;
+            _repeatCount = 0;
+        }
+        else if (ch == L';' && _repeatPending)
+        {
+            success = _applyPendingRepeat();
+        }
+        else
+        {
+            success = false;
+        }
+        break;
+    case State::ExpectingSecondHexDigit:
+        success = _decodeHexDigit(ch) && _appendToActiveMacro(_decodedChar);
+        _decodedChar = 0;
+        _parseState = State::ExpectingHexDigit;
+        break;
+    case State::ExpectingRepeatCount:
+        if (ch >= L'0' && ch <= L'9')
+        {
+            _repeatCount = _repeatCount * 10 + (ch - L'0');
+            _repeatCount = std::min<size_t>(_repeatCount, MAX_PARAMETER_VALUE);
+        }
+        else if (ch == L';')
+        {
+            _repeatPending = true;
+            _repeatStart = _activeMacro().length();
+            _parseState = State::ExpectingHexDigit;
+        }
+        else
+        {
+            success = false;
+        }
+        break;
+    default:
+        success = false;
+        break;
+    }
+
+    // If there is an error in the definition, clear everything received so far.
+    if (!success)
+    {
+        _deleteMacro(_activeMacro());
+    }
+    return success;
+}
+
+bool MacroBuffer::_decodeHexDigit(const wchar_t ch) noexcept
+{
+    _decodedChar <<= 4;
+    if (ch >= L'0' && ch <= L'9')
+    {
+        _decodedChar += (ch - L'0');
+        return true;
+    }
+    else if (ch >= L'A' && ch <= L'F')
+    {
+        _decodedChar += (ch - L'A' + 10);
+        return true;
+    }
+    else if (ch >= L'a' && ch <= L'f')
+    {
+        _decodedChar += (ch - L'a' + 10);
+        return true;
+    }
+    return false;
+}
+
+bool MacroBuffer::_appendToActiveMacro(const wchar_t ch)
+{
+    if (GetSpaceAvailable() > 0)
+    {
+        _activeMacro().push_back(ch);
+        _spaceUsed++;
+        return true;
+    }
+    return false;
+}
+
+std::wstring& MacroBuffer::_activeMacro() noexcept
+{
+    return til::at(_macros, _activeMacroId);
+}
+
+void MacroBuffer::_deleteMacro(std::wstring& macro) noexcept
+{
+    _spaceUsed -= macro.length();
+    std::wstring{}.swap(macro);
+}
+
+bool MacroBuffer::_applyPendingRepeat()
+{
+    auto& activeMacro = _activeMacro();
+    const auto sequenceLength = activeMacro.length() - _repeatStart;
+    if (sequenceLength && _repeatCount > 1)

```

Comment:
`sequenceLength` is the length of the text that is being repeated, right? Is the `if (sequenceLength)` check necessary at all then? From what I can tell, this code would work even without it and make it a bit easier to understand IMO, because then a reader only needs to focus on the `_repeatCount` parameter. For instance, I personally feel more concerned about the possibility of `_repeatStart` being greater than `activeMacro.length()`, which would underflow and not get catched by this if condition either.

Reply:
Checking the `sequenceLength` is just an optimisation. If it's zero and `_repeatCount` is large, it won't fail the space overflow test, so it'll end up sitting in massive loop repeatedly adding a blank string to the active macro. Perhaps that's not that big a deal, but why go through all that when it can easily be skipped.

And `_repeatStart` should never be greater than `activeMacro.length()`, but we can also double check that here if you prefer.

Reply:
> Checking the `sequenceLength` is just an optimisation. [...]

But do you think anyone would realistically emit a repeat sequence without anything that can be repeated? Even if someone did, would that take any longer than processing a non-empty repeat, which I'm sure is a far more common case? You can leave the code as is, it's just a "weak feeling" I had about this code. It took me a bit to fully grasp it initially, even if in hindsight it's now easy to understand and I was looking for ways to maybe make it simpler.

> And `_repeatStart` should never be greater than `activeMacro.length()`, but we can also double check that here if you prefer.

I'll leave that up to you. We have both checked and unchecked arithmetic in our code base, and it's hard to say whether this code deserves a checked one... I'd be 100% fine leaving it as is.

Reply:
> But do you think anyone would realistically emit a repeat sequence without anything that can be repeated? Even if someone did, would that take any longer than processing a non-empty repeat, which I'm sure is a far more common case?

I was thinking more of someone deliberately trying to do something malicious. Every terminal with macro support that I've tested could be crashed in some way, or at least forced to hang, so I've been trying to make sure that couldn't happen to us. But as you say, this wouldn't be any worse than the non-empty repeat, and the repeat count is already capped at `MAX_PARAMETER_VALUE`, so it's probably OK. I'll drop the check.